pipeline/top_level.py

```diff
@@ -82,45 +112,173 @@ def __init__(self, **kwargs):
+            else:
+                self.pars.override({key: value})
+
+    def augment_parameters(self, **kwargs):
+        """Add some parameters to this object and its sub-objects, using Parameters.augment(). """
+        for key, value in kwargs.items():
+            if key in PROCESS_OBJECTS:
+                getattr(self, PROCESS_OBJECTS[key]).pars.augment(value)
+            else:
+                self.pars.augment({key: value})
+
+    def setup_datastore(self, *args, **kwargs):
+        """Initialize a datastore, including an exposure and a report, to use in the pipeline run.
+
+        Will raise an exception if there is no valid Exposure,
+        if there's no reference available, or if the report cannot
+        be posted to the database.
+
+        After these objects are instantiated, the pipeline will proceed
+        and record any exceptions into the report object before raising them.
+
+        Parameters
+        ----------
+        Inputs should include the exposure and section_id, or a datastore
+        with these things already loaded. If a session is passed in as
+        one of the arguments, it will be used as a single session for
+        running the entire pipeline (instead of opening and closing
+        sessions where needed).
+
+        Returns
+        -------
+        ds : DataStore
+            The DataStore object that was created or loaded.
+        session: sqlalchemy.orm.session.Session
+            An optional session. If not given, this will be None
+        """
+        ds, session = DataStore.from_args(*args, **kwargs)
+
+        if ds.exposure is None:
+            raise RuntimeError('Not sure if there is a way to run this pipeline method without an exposure!')
+
+        try:  # must make sure the exposure is on the DB
+            ds.exposure = ds.exposure.merge_concurrent(session=session)
+        except Exception as e:
+            raise RuntimeError('Failed to merge the exposure into the session!') from e
+
+        try:  # make sure we have a reference for this exposure
+            with SmartSession(session) as dbsession:
+                reference = Reference.get_reference(ds.exposure.filter, ds.exposure.target, ds.exposure.observation_time, dbsession)
+                if reference is None:
+                    raise RuntimeError('No reference found!')
+        except Exception as e:
+            raise RuntimeError(f'Cannot get reference exposure {ds.exposure.filepath}!') from e
+
+        try:  # create (and commit, if not existing) all provenances for the products
+            with SmartSession(session) as dbsession:
+                provs = self.make_provenance_tree(ds.exposure, session=dbsession, commit=True)
+        except Exception as e:
+            raise RuntimeError('Failed to create the provenance tree!') from e
+
+        try:  # must make sure the report is on the DB
+            report = Report(exposure=ds.exposure, section_id=ds.section_id)
```

Comment:
So, if I understand correctly, every time we run the top level pipeline, a new report gets created, yes?

(This sounds like a reasonable way to do it.)

Reply:
yes.